keps/sig-node/4381-dra-semantic-parameters/README.md

```diff
@@ -0,0 +1,1477 @@
+    ...
+```
+
+Here all resources associated with the `cards.dra.example.com` driver are to
+be placed under the `partitionable` field in the `NodeResourceSlice` object. As
+we add more "semantic models" in the future, *alternate* fields will be added
+at the same level as `partitionable`, and driver implementors will be able to
+choose which one they want to use to represent their resources.
+
+Below is an example of a driver that provides two discrete GPU cards using the
+`partitionable` model. The first card can only be used as a whole. The second
+card can be used as a whole, split up into four equal pieces, in halves, or one
+half and two quarters:
+
+```yaml
+kind: NodeResourceSlice
+apiVersion: resource.k8s.io/v1alpha2
+...
+spec:
+  nodeName: worker-1
+  driverName: cards.dra.example.com
+  partitionable:
+    commonAttributes:
+    - name: drivers
+      attributes:
+      - name: driverVersion
+        String: 1.2.3
+      - name: runtimeVersion
+        String: 11.1.42
+    - name: t1000-gpu
+      attributes:
+      - name: type
+        string: GPU
+      - name: memory
+        quantity: 16Gi
+      - name: productName:
+        string: ACME T1000 32GB
+    - name: a4-gpu
+      attributes:
+      - name: type
+        String: GPU
+      - name: memory
+        quantity: 32Gi
+      - name: productName:
+        String: ACME A4-PCIE-40GB
+    - name: a4-half
+      attributes:
+      - name: memory
+        quantity: 16Gi
+    - name: a4-quarter
+      attributes:
+      - name: memory
+        quantity: 8Gi
+
+    resourceInstances:
+    - name: gpu-0
+      commonAttributesRefs:
+      - drivers
+      - t1000-gpu
+      attributes:
+      - name: UUID
+        string: GPU-ceea231c-4257-7af7-6726-efcb8fc2ace9
+    - name: partitionable-gpu-1
+      commonAttributesRefs:
+      - drivers
+      - a4-gpu
+      attributes:
+      - name: UUID
+        string: GPU-6aa0af9e-a2be-88c8-d2b3-2240d25318d7
+      partitions:
+      - name: full-gpu
+        resourceInstances:
+        - name: gpu-1
+          # Same attributes as referenced under partitionable-gpu-2.
+      - name: quarters
+        commonAttributesRefs:
+        - a4-quarter # less memory for each sub-instance.
+        resourceInstances:
+        - name: quarter-0
+        - name: quarter-1
+        - name: quarter-2
+        - name: quarter-3
+      - name: half-and-quarters
+        resourceInstances:
+        - name: first-half
+          comonAttributesRefs:
+          - a4-half
+        - name: partitionable-second-half
+          partitions:
+          - name: half
+            comonAttributesRefs:
+            - a4-half
+            resourceInstances:
+            - name: second-half
+          - name: quarters
+            comonAttributesRefs:
+            - a4-quarters
+            resourceInstances:
+            - name: quarter-0
+            - name: quarter-1
+```
+
+As mentioned previously, exactly one semantic model must be used by a given
```

Comment:
I still find this limiting, but I think that's non-blocking for this KEP. I can always add a composition model if I need it.

Similarly, I still think we want to push some things up a level. But I don't think it has to happen in this KEP. I would rather we get the basic machinery right in 1.30, and we can prototype and evaluate some additional models before we get to the 1.31 KEP freeze.

So, in short, while I still don't agree on all the details of the model, I consider that non-blocking.

Reply:
Agreed. As long as that experimentation is limited to one semantic model, the impact on the rest is limited. We could add "AlphaFoo" pointers to the one-of-many structs and remove them gain after a release when we are done or rename it when the experiment was successful.